fendermint/ipc/src/pof.rs

```diff
@@ -0,0 +1,85 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+//! IPC proof of finality related functions
+
+use crate::Config;
+use crate::{BlockHeight, ParentViewProvider};
+use anyhow::anyhow;
+use ipc_sdk::cross::CrossMsg;
+use ipc_sdk::ValidatorSet;
+use serde::{Deserialize, Serialize};
+use std::sync::Arc;
+
+/// The proof for POF.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub struct IPCParentFinality {
+    /// The latest chain height
+    pub height: u64,
+    /// The block hash. For FVM, it is a Cid. For Evm, it is bytes32.
+    pub block_hash: Vec<u8>,
+    /// new top-down messages finalized in this PoF
+    pub top_down_msgs: Vec<CrossMsg>,
+    /// latest configuration information from the parent.
+    pub config: ValidatorSet,
+}
+
+/// The parent finality information provider. It provides the latest finality information about the
+/// parent. Also it perform validation on incoming finality.
+#[derive(Clone)]
+pub struct ParentFinalityProvider<T> {
+    config: Config,
+    parent_view_provider: Arc<T>,
+    latest_confirmed_finality: Option<IPCParentFinality>,
+}
+
+impl<T: ParentViewProvider> ParentFinalityProvider<T> {
+    pub fn finality_committed(&mut self, finality: IPCParentFinality) -> anyhow::Result<()> {
+        self.parent_view_provider.on_finality_committed(&finality);
+        self.latest_confirmed_finality.replace(finality);
+        Ok(())
+    }
+
+    /// The next finality to propose
+    pub fn next_finality_proposal(&self) -> anyhow::Result<Option<IPCParentFinality>> {
+        let latest_height = match self.parent_view_provider.latest_height() {
+            Some(h) => h,
+            None => return Ok(None),
+        };
+        if latest_height < self.config.chain_head_delay {
+            return Ok(None);
+        }
+
+        let confident_height = latest_height - self.config.chain_head_delay;
+        self.finality_proposal_at_height(confident_height)
```

Comment:
It looks like we never access the data in anything above `latest_height - chain_head_delay`, so I wonder why the syncer even fetches those? It could just ask the parent chain always about the level which it thinks is final, and save itself the trouble of querying the IPC Agent at every poll interval for the last 900 or heights, only to look at the 900th and nothing else.

Reply:
It's for preparing the data in advance. This could be queried when `process_proposal`, someone else sends a height that is greater than `latest_height - chain_head_delay`, we still need to fetch in advance to get the data so that we can validate.